### ISSUE TITLE
[rfc] stabilised wire output for proselint

### ISSUE THREAD

**Issue by Nytelife26** [amperser/proselint#1427]
## Relevant issues

Follows on from discussions in #1307.

CC: @jclicodes @drainpixie 

## Abstract

Currently, proselint's wire mode (JSON) is unstable, and not guaranteed to produce valid output in the target format. This proposal will resolve this by introducing a standard schema for proselint's wire mode.

Credit to @jclicodes for streamlining the proposed interface.

## Rundown

This section contains a model of proselint's current output in wire mode. You may skip it if you are already familiar with proselint's behaviour.

```ts
interface CheckResult {
    check_path: string;
    message: string;
    start_pos: number;
    end_pos: number;
    replacements?: string;
}

interface LintResult extends CheckResult {
    line: number;
    column: number;
    severity: string;
}

interface ProselintOutput {
    data: { errors: LintResult[] };
    status: "success";
}
```

The `status` key is always `success`, because if proselint errors, the output will not be valid for the specified wire format. This is highly undesirable behaviour for external communications, and [creates errors upstream] in certain situations. `data.errors` is a list of all the warnings generated by proselint over the file. The top-level model `ProselintOutput` is produced once for each file.

[creates errors upstream]: github.com/amperser/proselint/issues/1307

## Proposal

Eventually, it would be great to provide an implementation of the Language Server Protocol. For now, we will set the following goals:

1. Always produce valid wire output, even when proselint encounters an error
2. Produce one model *per-run*, rather than per-file

The wire output will be versioned, and have its version tied to that of proselint itself.

### Infallible operation

To achieve infallibility, proselint will use mutually-exclusive keys with a separate schema for successful and failed operations.

### One output per run

To achieve a single output per-run, rather than per-file, the `data` record will be updated to support multiple keys. Each key will be the name of a lint file, and the success value will be an array of results returned from running
proselint over the file.

### Intended schema

Overall, the new response schema will work as follows, inheriting `LintResult` as previously defined.

```ts
interface ResponseError {
    code: number;
    message: string;
}

interface FileOutput {
    diagnostics?: LintResult[];
    error?: ResponseError;
}

interface ProselintOutput {
    result?: Record<string, FileOutput>;
    error?: ResponseError;
}
```

Both `ProselintOutput` and `FileOutput` use mutually-exclusive keys; that is, `error` is only present if `result` or `diagnostics` respectively are not, and vice versa. One of them must always be present in the output for each type. These two guarantees form a discriminated union, allowing end-users to quickly determine whether proselint failed over an individual file, or altogether. The use of a discriminated union here makes `status` redundant, so it has been removed.

The `result` and `error` objects have been named to correspond with their equivalents in the LSP specification. `error.code` will adhere to the reserved codes in the `ErrorCodes` namespace of the LSP specification, and this proposal will be updated with the proselint-specific codes when possible.

**Comment by Nytelife26** [comment]
My original thought was that the run either succeeds as a whole, or fails as a whole, but you are correct in saying that it would be useful to have output for files that the linter succeeded against even if some fail. In the case of `pylint`, failing to parse a module constitutes a lint result of its own, so they can get away with bundling this data with the rest of the lint results. For our purposes, do you think it might make more sense to allow `result and  `error` to coexist, and to change `error` to `Record<string, ResponseError`? That would represent a record of paths and the corresponding error proselint encountered trying to process them.

**Comment by jclicodes** [comment]
Pylint represents the error as a `LintResult` - as you can see it has line: 1 and column: 0 hard-coded (for example). This seems rather clumsy to me, plus it mixes semantics (internal errors arent linting failures so it doesnt make sense to represent them as one). 

Your solution separates the actual diagnostics useful to a third-party tool from the linter errors. However, in my opinion, this would complicate trying to use the tool for authors of third-party tools as they now have two places to look, and must merge two maps to know each file's state.

You could perhaps blend the two solutions.

```typescript
interface CheckResult {
    check_path: string;
    message: string;
    start_pos: number;
    end_pos: number;
    replacements?: string;
}

interface Diagnostic extends CheckResult {
    line: number;
    column: number;
    severity: string;
}

type LintResult = Diagnostic[] | ResponseError

interface ProselintOutput {
    result?: Record<string, LintResult>;
    error?: ResponseError;
}
```

This way we also reserve a mutually exclusive top-level `error` field for any truly run-level errors like config errors, invalid args, etc. And we avoid mixing semantics as per-file parsing errors aren't converted into linter warnings.

This also means third party tools (including potentially an LSP server) can parse through the one map.

Example output:

```json
{
  "result": {
    "path/ok.md": [
       { // ... }, { // ...} 
    ],
    "path/bad.md": {
      "code": "EACCES", "message": "Permission denied"
    }
  }
}

```

Personally I think this approach isn't exactly ready either. The idea that a developer might have to discern between a file IO issue versus a valid linter warning using the data type (array = diagnostic, map = error) is a bad idea because they aren't self-describing, consumers would have to write branching logic based on runtime types. 

Instead, you could go with a discriminated union:

```json
{
  "result": {
    "path/ok.md": {
       "diagnostics": [{ // ... }, { // ...}] 
    },
    "path/bad.md": {
       "error": {
           "code": "EACCES", "message": "Permission denied"
       }
    }
  }
}
```

That way the contract is more explicit. The type can be discerned from the discriminant key.

**Comment by Nytelife26** [comment]
I spent the past few days looking it over, and I have no reservations so far. In fact, it sounds simple enough that we could go straight for it without the intermediate representation originally planned for 1.0. I'll write up a new version of the proposal with your ideas implemented now.
